File: backend/app/schemas/backtest.py
from pydantic import BaseModel


class BacktestMetrics(BaseModel):
    n_predictions: int
    directional_accuracy: float
    baseline_accuracy: float
    persistence_accuracy: float
    mae: float
    rmse: float
    baseline_mae: float
    beats_baseline: bool


class BacktestOut(BaseModel):
    id: str
    ticker: str
    horizon: str
    model_version: str
    start_date: str | None = None
    end_date: str | None = None
    status: str
    metrics: BacktestMetrics
    chart_data: dict
    note: str
    disclaimer: str
    created_at: str | None = None


class BacktestSummary(BaseModel):
    id: str
    ticker: str
    horizon: str
    start_date: str | None = None
    end_date: str | None = None
    metrics: BacktestMetrics
    created_at: str | None = None

File: frontend/types/backtest.ts
import type { Horizon } from "@/types/prediction";

export type { Horizon };

export interface BacktestMetrics {
  n_predictions: number;
  directional_accuracy: number;
  baseline_accuracy: number;
  persistence_accuracy: number;
  mae: number;
  rmse: number;
  baseline_mae: number;
  beats_baseline: boolean;
}

export interface EquityPoint {
  date: string;
  model: number;
  buy_hold: number;
}

export interface Backtest {
  id: string;
  ticker: string;
  horizon: string;
  model_version: string;
  start_date: string | null;
  end_date: string | null;
  status: string;
  metrics: BacktestMetrics;
  chart_data: {
    equity: EquityPoint[];
    cumulative_accuracy: { date: string; accuracy: number }[];
    returns: { date: string; pred_return: number; actual_return: number }[];
  };
  note: string;
  disclaimer: string;
  created_at: string | null;
}
